packages/@uppy/companion-client/src/Provider.js

```diff
@@ -177,9 +210,10 @@ export default class Provider extends RequestClient {
 
       return await super.request(...args)
     } catch (err) {
+      if (!this.supportsRefreshToken) throw err
       // only handle auth errors (401 from provider), and only handle them if we have a (refresh) token
       const authTokenAfter = await this.#getAuthToken()
-      if (!err.isAuthError || !authTokenAfter) throw err
+      if (!(err instanceof AuthError) || !authTokenAfter) throw err
```

Comment:
Why use `instanceof`? It's usually more reliable to test for properties instead – e.b. in the case where there are for some reason 2 versions of the `AuthError` class

Reply:
I thought checking for a specific type is more reliable than duck typing. for example if there's some other (3rd party) code that throws a different kind of error which happens to have the same property set (in this case `isAuthError`) then it would be a false positive and we would accidentally treat it as our own AuthError.

Another reason for using instanceof is that once we refactor to typescript, we don't have to change anything, and it will just work. If we use err.isAuthError, we probably have to do something like this in typescript which is long and cumbersome:
```js
if (!(typeof err === 'object' && err != null && 'isAuthError' in err && typeof err.isAuthError === 'boolean' && err.isAuthError) || !authTokenAfter) throw err
```

Reply:
No TS is fine with `err?.isAuthError`, you don't need to check whether it's a boolean.

Reply:
Nah, adding runtime checks to please TS is not the way to go, here's how I would write it: `(err as AuthError)?.isAuthError`. You could also use `err != null && 'isAuthError' in err`, the other checks are not necessary.

> which is correct because ts doesn't know that that property exists

Well in this case, TS is being dumb, the fact that the property doesn't exist shouldn't matter.

Reply:
I would like to add there are also type predicates:
https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

You can easily implement an `isAuthError()` function that determines whether the error object is an auth error based on duck typing - and then typescript is fully aware of the type

```
type AuthError {
...
}

function isAuthError(err: Error): err is AuthError {
  return `isAuthError` in err;
}
```

Reply:
> ```js
> import {got, HTTPError} from 'got'
> try {
>   await got('transloadit.com');
> } catch (err) {
>   const httpError = err as HTTPError;
>   if (httpError?.response.statusCode && httpError?.response.requestUrl) {
>     console.log(httpError.response.statusCode, httpError?.response.requestUrl)
>   }
> }
> ```

That's kind of a straw man, I don't think you would ever write this in JS, so logically you wouldn't in TS either. And you are ignoring @dschmidt's suggestion for some reason, which would make the code way more logical:

```typescript
import {got, type HTTPError} from 'got'

function isAuthError(err?: Error): err is HTTPError {
  return err != null &&
         `response` in err &&
         (err as HTTPError).response?.statusCode &&
         !!err.response.requestUrl;
}

try {
  await got('transloadit.com')
} catch (err) {
  if (isHTTPError(err)) {
    console.log(err.response.statusCode, err.response.requestUrl)
  }
}
```

Anyway, if you don't think my concerns apply, feel free to ignore, and let's move on.

Reply:
> That's kind of a straw man, I don't think you would ever write this in JS, so logically you wouldn't in TS either

I used that example because I actually did write code exactly like that in JS, and I think porting it to TS requires minimal effort:
https://github.com/transloadit/uppy/blob/58869e62873440b6960d025cbc739ecbcf50c786/packages/%40uppy/companion/src/server/provider/providerErrors.js#L35

> And you are ignoring @dschmidt's suggestion for some reason, which would make the code way more logical:

Yes that works, but I feel it's very error prone and extra work because we have to manually make (and maintain) these helpers for every error class that we want to catch, just to please TS. If we forget to check for one property inside `isHTTPError` (or forgot one `?.`) then we have a crash waiting to happen (probably in production).

Your concern is about reliability, right? Do you have examples where using instanceof is unreliable?

Reply:
If you're getting an error from a different realm – or if there are two versions of the same error class for whatever reason – `instanceof` only catches the instances created from the class you referenced (assuming no one has touched  `Symbol.hasInstance`).

Reply:
i'm not familiar with realms. In Uppy when would an error be thrown/catch'ed across realms? also when would an error class have multiple versions? Are these reliability problems Uppy specific or in general? I haven't seen much advice against using `instanceof` in general when catching errors. I've tried to search around for how to catch errors in typescript, and there isn't any official advice from the typescript team, but it seems that virtually all examples out there and stackoverflows use `err instanceof Error`. and i cannot find any typescript best practices that advice against using instanceof, so I'm struggling to understand why "everyone" is using something if it's considered unsafe.

Reply:
> I'm struggling to understand why "everyone" is using something if it's considered unsafe.

You can say the same thing about using Express to setup a server with Node.js, and draw the same conclusion; or why “everyone” is still using fossil fuels if using them is a threat to our survival 🤷‍♂️ 

> Are these reliability problems Uppy specific or in general?

These are problems in general, certainly not Uppy specific. Whether or not such use case is realistic with Uppy, I don’t know. Realms are not yet stage 4 IIRC, so problems could arise in the future if that language feature gets traction. Anyway, it’s very much a theoretical risk. 

> I haven't seen much advice against using `instanceof` in general when catching errors

Jest has this limitation that authors cannot use `instanceof` in the tests, precisely because test and code are on separate realms. Not related to our use case, but I wanted to mention it as an example.

Reply:
If we agree to not use `instanceof`, maybe we should create a new PR to remove this usage, because we've used instanceof many places in our codebase.

Reply:
I mean yeah, we should aim for consistency eventually. But for this PR, it seems preferable to keep things as is since this is not related (I don't think)